src/goto-instrument/contracts/contracts.cpp

```diff
@@ -331,23 +381,11 @@ void code_contractst::check_apply_loop_contracts(
   havoc_gen.append_full_havoc_code(
     loop_head->source_location(), generated_code);
 
-  // Add the havocing code, but only check against the enclosing scope's
-  // write set if it was manually specified.
-  if(assigns_clause.is_nil())
-    insert_before_swap_and_advance(
-      goto_function.body,
-      loop_head,
-      add_pragma_disable_assigns_check(generated_code));
-  else
-    insert_before_swap_and_advance(
-      goto_function.body, loop_head, generated_code);
-
-  // ^^^ FIXME ^^^
-  // comment by delmasrd: I did not reactivate this behaviour
-  // because I think we always want to check the loop assignments against
-  // the surrounding clause
-
-  insert_before_swap_and_advance(goto_function.body, loop_head, generated_code);
+  // Insert the havocing code.
+  // We do not `add_pragma_disable_assigns_check`,
+  // so that the enclosing scope's assigns clause instrumentation
+  // would pick these up for inclusion (subset) checks.
+  goto_function.body.destructive_insert(loop_head, generated_code);
```

Comment:
could we use separate variables for the different insertion steps ? This would make it clearer what gets inserted where.

Reply:
Yeah that's a good point. I will make this change. It was simple in the beginning, but now that we have 3 different contract types it's quite messy :D